File: Php/core/Validation/BaseValidator.php
<?php
namespace core\Validation{

	abstract class BaseValidator implements Validator{

		protected $errors;

		public function addError($message){
			$this->ensureErrorArray();
			$this->errors[] = $message;
		}

		public function getErrors(){
			$this->ensureErrorArray();
			return $this->errors;
		}

		protected function ensureErrorArray(){
			if($this->errors == null){
				$this->errors = array();
			}
		}

		abstract protected function validate();

	}

}

File: Web/Validators/BaseValidator.cs
using System.Collections.Generic;

namespace Web.Validators{

	public abstract class BaseValidator : Validator{

        protected List<string> Errors;

		public void AddError(string message){
			EnsureErrorArray();
            Errors.Add(message);
		}

		public virtual List<string> GetErrors(){
			EnsureErrorArray();
			return Errors;
		}

		protected void EnsureErrorArray(){
			if(Errors == null){
				Errors = new List<string>();
			}
		}

		abstract protected void Validate();

        public abstract bool IsValid();

	}

}